File: python_package/src/ginormal/dgin.py
import numpy as np
import warnings
from ginormal.utils.dgin1 import dgin1

def dgin(z, alpha, mu, tau, log=True, quasi=False):
    """
    Density for the generalized inverse normal distribution.

    Parameters
    ----------
    z : real number
        Distribution is supported on real line, z can be any real number.
        Currently, only scalar z is supported. Density is set to be 0 at z = 0.
    alpha : real number > 1
        Degrees of freedom parameter. Function returns 0 density if alpha <= 1.
    mu : real number
        Similar to location parameter, controls asymmetry of the distribution.
    tau : real number > 0
        Similar to scale parameter, controls spread of the distribution.
    log : boolean, optional
        Should the log of the density be returned? Default is True.
    quasi : boolean, optional
        The quasi-density or kernel is the density without the normalization constant.
        Should the quasi-density value be returned? Default is False.
    """
    # Check parameter values
    if alpha <= 1:
        warnings.warn("alpha should be greater than 1")
        res = -np.inf
    elif tau <= 0:
        warnings.warn("tau should be greater than 0")
        res = -np.inf
    elif z == 0:
        res = -np.inf
    else:
        # Compute density
        mt = mu / tau
        if quasi:
            res = dgin1(z * tau, alpha, mt, True, True) + alpha * np.log(tau)
        else:
            res = dgin1(z * tau, alpha, mt, True, False) + np.log(tau)
    
    # Return
    if log:
        return res
    else:
        return np.exp(res)

File: ginormal/R/dgin.R
#' Density for the generalized inverse normal distribution
#'
#' @importFrom stats runif
#' @param z quantile.
#' @param alpha degrees-of-freedom parameter.
#' @param mu similar to location parameter, controls asymmetry of the distribution.
#' @param tau similar to scale parameter, controls spread of the distribution.
#' @param log logical; should the log of the density be returned? Defaults to TRUE.
#' @param quasi logical; should the quasi-density value be returned? Defaults to FALSE.
#' @details
#' Currently, only scalars are supported for the quantile and parameter values.
#' Density is supported on the entire real line, `z` and `mu` can take any value
#' in $(-\\infty, \\infty)$. Density is only defined for parameter values
#' `alpha`${} > 1$ or `tau`${} > 0$, so it is set to 0 outside of these values.
#' The quasi-density or kernel is the density without the normalization constant,
#' use `quasi = TRUE` for this behavior.
#'
#' @return Numeric scalar with density.
#' @export dgin
#'
#' @examples
#' # Computing (log) density
#' dgin(z = 1, alpha = 3, mu = 1, tau = 1, log = TRUE, quasi = FALSE)
#'
#' # Generalized inverse normal density with alpha = 5, mu = 0, tau = 1
#' z_vals <- seq(-5, 5, length.out = 200)
#' fz_unc <- sapply(z_vals, function(z) dgin(z, 5, 0, 1, FALSE))
#' plot(z_vals, fz_unc, type = "l", xlab = 'Values', ylab = 'Density')
dgin <- function(z, alpha, mu, tau, log=TRUE, quasi=FALSE){
  # Check parameter values
  if (alpha <= 1) {
    warning("alpha should be greater than 1")
    res <- -Inf
  } else if (tau <= 0) {
    warning("tau should be greater than 0")
    res <- -Inf
  } else if (z == 0) {
    res <- -Inf
  } else {
    # Compute density
    mt <- mu / tau
    if (quasi == TRUE){
      res <- dgin1(z * tau, alpha, mt, TRUE, TRUE) + alpha * log(tau)
    } else {
      res <- dgin1(z * tau, alpha, mt, TRUE, FALSE) + log(tau)
    }
  }

  # Return
  if (log == TRUE){
    return(res)
  } else {
    return(exp(res))
  }
}
